src/server/test/hotkey_collector_test.cpp

```diff
@@ -195,5 +198,174 @@ TEST_F(fine_collector_test, fine_collector)
+                     resp);
+    ASSERT_EQ(resp.err, dsn::ERR_OK);
+    ASSERT_EQ(get_collector_stat(collector), hotkey_collector_state::COARSE_DETECTING);
+
+    for (int i = 0; i < 100; i++) {
+        dsn::tasking::enqueue(LPC_WRITE, &_tracker, [&] {
+            _server->on_get(generate_get_rpc(generate_hash_key_by_random(true, 80)));
+        });
+    }
+    _tracker.wait_outstanding_tasks();
+    collector->analyse_data();
+    ASSERT_EQ(get_collector_stat(collector), hotkey_collector_state::FINE_DETECTING);
+
+    for (int i = 0; i < 100; i++) {
+        dsn::tasking::enqueue(LPC_WRITE, &_tracker, [&] {
+            _server->on_get(generate_get_rpc(generate_hash_key_by_random(true, 80)));
+        });
+    }
+    _tracker.wait_outstanding_tasks();
+    collector->analyse_data();
+    ASSERT_EQ(get_collector_stat(collector), hotkey_collector_state::FINISHED);
+
+    auto result = get_result(collector);
+    ASSERT_TRUE(result->if_find_result);
+    ASSERT_EQ(result->hot_hash_key, "ThisisahotkeyThisisahotkey");
+
+    on_detect_hotkey(generate_control_rpc(dsn::replication::hotkey_type::READ,
+                                          dsn::replication::detect_action::STOP),
+                     resp);
+    ASSERT_EQ(resp.err, dsn::ERR_OK);
+    ASSERT_EQ(get_collector_stat(collector), hotkey_collector_state::STOPPED);
+
+    on_detect_hotkey(generate_control_rpc(dsn::replication::hotkey_type::READ,
+                                          dsn::replication::detect_action::START),
+                     resp);
+    ASSERT_EQ(resp.err, dsn::ERR_OK);
+    ASSERT_EQ(get_collector_stat(collector), hotkey_collector_state::COARSE_DETECTING);
+
+    for (int i = 0; i < 1000; i++) {
+        dsn::tasking::enqueue(LPC_WRITE, &_tracker, [&] {
+            _server->on_get(generate_get_rpc(generate_hash_key_by_random(false)));
+        });
+    }
+    collector->analyse_data();
+    ASSERT_EQ(get_collector_stat(collector), hotkey_collector_state::COARSE_DETECTING);
+
+    on_detect_hotkey(generate_control_rpc(dsn::replication::hotkey_type::READ,
+                                          dsn::replication::detect_action::STOP),
+                     resp);
+    ASSERT_EQ(resp.err, dsn::ERR_OK);
+    ASSERT_EQ(get_collector_stat(collector), hotkey_collector_state::STOPPED);
+    _tracker.wait_outstanding_tasks();
+}
+
+TEST_F(hotkey_collector_test, data_completeness)
+{
+    dsn::replication::detect_hotkey_response resp;
+    on_detect_hotkey(generate_control_rpc(dsn::replication::hotkey_type::READ,
+                                          dsn::replication::detect_action::START),
+                     resp);
+    ASSERT_EQ(resp.err, dsn::ERR_OK);
+    on_detect_hotkey(generate_control_rpc(dsn::replication::hotkey_type::WRITE,
+                                          dsn::replication::detect_action::START),
+                     resp);
+    ASSERT_EQ(resp.err, dsn::ERR_OK);
+
+    auto writes = new dsn::message_ex *[1000];
```

Comment:
Use message_ptr here. This line has definitely memory leak.

Reply:
`auto cleanup = dsn::defer([=]() { delete[] writes; });`
I defer deleting this object, is it not correct?

Reply:
`delete[] writes` only deletes the array, but not deletes the elements. Use STL& smart ptr whenever possible.